design/httpproxy-prefix-rewrite.md

```diff
@@ -0,0 +1,262 @@
+
+HTTPProxy documents are organized into a collection of trees, where the root is a document that has a `virtualhost` field, and the leaves are documents that have a `routes` field.
+A HTTPProxy tree could be a single document (that is both a root and a leaf), or it could be a series of documents in different namespaces, connected by the `includes` field.
+
+HTTPProxy inclusion semantics allow a single HTTPProxy CRD to be included by more than one root HTTPProxy document.
+This means that multiple routing path prefixes can end up referring to the same HTTPProxy.
+
+To reason about most kinds of request rewriting, the user needs to know the final form of the request.
+In other words, it doesn't make sense to want to rewrite a path prefix unless you have some expectations about the path prefix you will receive.
+This implies that even if the owner of a leaf HTTPProxy doesn't control the root or intermediate documents, they still have fairly concrete expectations around the full URL path to expect.
+
+Although there are a number of ways to conceptualize path prefix rewriting, path regex rewriting requires knowledge of the full request path at the time of writing the regex.
+This is because a regex rewrite can be used to re-order arbitratry portions of the URL path, and this can't be done unless the regex is defined as being applied to the complete path.
+
+## High Level Design
+
+Add a `PathRewrite` field to the HTTPProxy [Route](https://github.com/projectcontour/contour/blob/master/apis/projectcontour/v1/httpproxy.go#L125) object.
+This allows us to specify different kinds of rewrite for URL paths without also needing to specify a specific kind of request rewriting (i.e. we can add host rewrite later).
+
+Locating the `PathRewritePolicy` on the `Route` means that rewriting is controlled by the team that controls the leaf HTTPProxy document.
+This allows a useful initial implementation, but does not prevent adding `PathRewritePolicy` fields higher up the HTTPProxy tree (though allowing that would require detecting and resolving conflicts).
+
+```Go
+type Route struct 
+	...
+
+	// The policy for rewriting the path of the request URL
+	// after the request has been routed to a Service.
+	PathRewrite *PathRewritePolicy `json:"pathRewrite,omitempty"`
+}
+
+```
+
+The `PathRewritePolicy` struct initially only supports path prefix replacement.
+Creating a policy struct allows future addition of other kinds of path rewriting.
+
+
+```Go
+// PathRewritePolicy specifies how a request URL path should be
+// rewritten. This rewriting takes place after a request is routed
+// and has no subsequent effects on the proxy's routing decision.
+// 
+// Exactly one field in this struct may be specified.
+type PathRewritePolicy struct {
+    ReplacePrefix []struct {
+        // +optional
+        Prefix string `json:"prefix,omitempty"`
```

Comment:
Could you add doc comments to these fields please

Reply:
For now, the fields are described in the following text. If we add this to the code, that text would move to docs comments.

Reply:
I'd really appreciate it if you could summarise the use of those fields in this design document.

The reason I ask this is I want to see the use of `Prefix`. Its not clear to me what this field does as the matching of the containing route is not part of the `Prefix` field. Making this field optional makes me wonder under which circumstances it is needed. I think a doc comment right here in the code sample would help clarify these questions for me.